broker/cloud_functions/lsst/ps_to_gcs/main.py

```diff
@@ -0,0 +1,196 @@
+#!/usr/bin/env python3
+# -*- coding: UTF-8 -*-
+
+"""This module stores the alert data as an Avro file in Cloud Storage,
+fixing the schema first, if necessary.
+"""
+
+import base64
+import io
+import json
+import os
+import struct
+from typing import Optional
+
+import fastavro
+import pittgoogle
+from confluent_kafka.schema_registry import SchemaRegistryClient
+from google.cloud import functions_v1, logging, storage, pubsub_v1
+from google.cloud.exceptions import PreconditionFailed
+
+
+PROJECT_ID = os.getenv("GCP_PROJECT")
+TESTID = os.getenv("TESTID")
+SURVEY = os.getenv("SURVEY")
+VERSIONTAG = os.getenv("VERSIONTAG")
+
+# connect to the cloud logger and publisher
+logging_client = logging.Client()
+log_name = "ps-to-gcs-cloudfnc"
+logger = logging_client.logger(log_name)
+publisher = pubsub_v1.PublisherClient()
+
+# GCP resources used in this module
+client = storage.Client()
+available_schemas = {
+    "7.1": "v7_1",
+    "7.2": "v7_2",
+    "7.3": "v7_3",
+}
+ALERTS_TOPIC = pittgoogle.Topic.from_cloud(
+    "alerts", survey=SURVEY, testid=TESTID, projectid=PROJECT_ID
+)
+BUCKETS = {}
+
+# alerts are stored in GCS buckets based on their schema version; alerts in topic may contain multiple schema versions.
+# to avoid making the get_bucket call for each alert, we'll cache the buckets and assign the correct bucket dynamically
+# based on the alert's schema version
+for versiontag in available_schemas.values():
+    bucket_name = f"{PROJECT_ID}-{SURVEY}_alerts_{versiontag}"
+    if TESTID != "False":
+        bucket_name = f"{bucket_name}-{TESTID}"
+    BUCKETS[versiontag] = client.get_bucket(client.bucket(bucket_name, user_project=PROJECT_ID))
```

Comment:
Pre-defining the buckets and making a .get_bucket request for each instance rather than each alert is a temporary solution I discussed with @troyraen. For future schema implementations, this will need to be updated